client/options.go

```diff
@@ -131,239 +130,542 @@ func (cliOpts *ClientOptions) init(opts ...ClientOption) error {
+	return func(opts *ReferenceOptions) {
+		opts.Reference.Loadbalance = constant.LoadBalanceKeyConsistentHashing
 	}
 }
 
-func WithShutdown(opts ...graceful_shutdown.Option) ClientOption {
-	sdOpts := graceful_shutdown.NewOptions(opts...)
+func WithLoadBalanceLeastActive() ReferenceOption {
+	return func(opts *ReferenceOptions) {
+		opts.Reference.Loadbalance = constant.LoadBalanceKeyLeastActive
+	}
+}
+
+func WithLoadBalanceRandom() ReferenceOption {
+	return func(opts *ReferenceOptions) {
+		opts.Reference.Loadbalance = constant.LoadBalanceKeyRandom
+	}
+}
+
+func WithLoadBalanceRoundRobin() ReferenceOption {
+	return func(opts *ReferenceOptions) {
+		opts.Reference.Loadbalance = constant.LoadBalanceKeyRoundRobin
+	}
+}
+
+func WithLoadBalanceP2C() ReferenceOption {
+	return func(opts *ReferenceOptions) {
+		opts.Reference.Loadbalance = constant.LoadBalanceKeyP2C
+	}
+}
+
+func WithLoadBalanceXDSRingHash() ReferenceOption {
+	return func(opts *ReferenceOptions) {
+		opts.Reference.Loadbalance = constant.LoadBalanceKeyLeastActive
+	}
+}
+
+func WithRetries(retries int) ReferenceOption {
+	return func(opts *ReferenceOptions) {
+		opts.Reference.Retries = strconv.Itoa(retries)
+	}
+}
+
+func WithGroup(group string) ReferenceOption {
+	return func(opts *ReferenceOptions) {
+		opts.Reference.Group = group
+	}
+}
+
+func WithVersion(version string) ReferenceOption {
+	return func(opts *ReferenceOptions) {
+		opts.Reference.Version = version
+	}
+}
+
+func WithJSON() ReferenceOption {
+	return func(opts *ReferenceOptions) {
+		opts.Reference.Serialization = constant.JSONSerialization
+	}
+}
+
+func WithProvidedBy(providedBy string) ReferenceOption {
+	return func(opts *ReferenceOptions) {
+		opts.Reference.ProvidedBy = providedBy
+	}
+}
+
+// todo(DMwangnima): implement this functionality
+//func WithAsync() ReferenceOption {
+//	return func(opts *ReferenceOptions) {
+//		opts.Reference.Async = true
+//	}
+//}
+
+func WithParams(params map[string]string) ReferenceOption {
+	return func(opts *ReferenceOptions) {
+		opts.Reference.Params = params
+	}
+}
+
+// todo(DMwangnima): implement this functionality
+//func WithGeneric(generic bool) ReferenceOption {
+//	return func(opts *ReferenceOptions) {
+//		if generic {
+//			opts.Reference.Generic = "true"
+//		} else {
+//			opts.Reference.Generic = "false"
+//		}
+//	}
+//}
+
+func WithSticky(sticky bool) ReferenceOption {
+	return func(opts *ReferenceOptions) {
+		opts.Reference.Sticky = sticky
+	}
+}
+
+// ========== Protocol to consume ==========
+
+func WithProtocolDubbo() ReferenceOption {
+	return func(opts *ReferenceOptions) {
+		opts.Reference.Protocol = constant.Dubbo
+	}
+}
+
+func WithProtocolTriple() ReferenceOption {
+	return func(opts *ReferenceOptions) {
+		opts.Reference.Protocol = "tri"
+	}
+}
+
+func WithProtocolJsonRPC() ReferenceOption {
+	return func(opts *ReferenceOptions) {
+		opts.Reference.Protocol = "jsonrpc"
+	}
+}
+
+func WithProtocol(protocol string) ReferenceOption {
```

Comment:
Since we have enumerated protocols that we support, maybe there is no need to expose this API?

Reply:
What if the user custom protocol?

Reply:
Yeah, I think we should provide both for flexibility.